soco/core.py

```diff
@@ -1833,6 +1872,181 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
             ('AddAtIndex', 4294967295)
         ])
 
+    def reorder_sonos_playlist(self, object_id, tracks, new_pos, update_id=0):
+        """ Reorder and/or Remove tracks in a Sonos playlist. The underlying
+        call is quite complex. See the helper methods
+        :py:meth:`clear_sonos_playlist`, :py:meth:`move_in_sonos_playlist`,
+        :py:meth:`remove_from_sonos_playlist` for simplified usage.
+
+        update_id - If you have a series of operations, tracking the update_id
```

Comment:
I've expanded the doc string some what.  But I'll explain in more detail
here and based on feedback, I'll remodify the docstring.

The call can reorder and delete tracks.  In general the underlying call
expects two string arguments, tracks which indicate the 'current' track
position you want to modify and the new_pos that dictate what should happen
to it.  You can, in a single request, specify moving and deleting, but they
happen grouped into two calls.  First doing all the delete requests in an
op, and then doing the moves (however that is with respect to the new track
positions after the deletes.  If you delete a track, all following tracks
are immediately renumbered to fill in the gap. So any move requests need to
take this into account.  It is simpler to think about if you make the
changes individually.

So the first option of list of ints.   tracks=[1,3], new_pos=[0, None]
(since it is using ints, I chose to indicate a delete with None) The
results of this operation would move the second track into the first
position, what was in the 1 position moves to the second postion and the
4th track is deleted from the playlist.

trackA       pos:0
trackB       pos:1
trackC       pos:2
trackD       pos:3

is transformed into:
trackB      pos:0
trackA      pos:1
trackC      pos:2

When the inputs are lists, each pair zip(tracks, new_pos) creates an
underlying call to avTransport.ReorderTracksInSavedQueue.

The same call using a list of int-likes  would look like:  tracks=['1',
'3'], new_pos['0', ''], which would also create two underlying calls.
(note: when using int-likes a delete is specified by an empty string.)

Now, if you really wanted to have fun, you can call reorder with strings,
tracks='1,3,5', new_pos='0,2,4' which would move the 2nd, 4th and 6th track
into the first, third and 5th position in a single underlying call. Using a
string, says: "I know what I'm doing, get out of my way."  It seems to
happen "atomically"  in that the moves happen before the track order
shifts.  However, getting this right can be frustrating when you have
deletes going on at the same time.   Using the helper functions clear, move
and remove make it simpler to think about because the list is being
transformed one step at a time and you only need to supply simple parameters

maybe we should prepend an underscore to reorder_sonos_playlist ->
_reorder_sonos_playlist. To encourage people to use just the helpers. ??

On Mon, Sep 7, 2015 at 10:48 AM, Lawrence Akka <notifications@github.com>
wrote:

> In soco/core.py
> <https://github.com/SoCo/SoCo/pull/352#discussion_r38870720>:
>
> > @@ -1833,6 +1872,181 @@ def add_item_to_sonos_playlist(self, queueable_item, sonos_playlist):
> >              ('AddAtIndex', 4294967295)
> >          ])
> >
> > +    def reorder_sonos_playlist(self, object_id, tracks, new_pos, update_id=0):
> > +        """ Reorder and/or Remove tracks in a Sonos playlist. The underlying
> > +        call is quite complex. See the helper methods
> > +        :py:meth:`clear_sonos_playlist`, :py:meth:`move_in_sonos_playlist`,
> > +        :py:meth:`remove_from_sonos_playlist` for simplified usage.
> > +
> > +        update_id - If you have a series of operations, tracking the update_id
>
> Please can you explain more - if not in the doctstring then in a normal
> comment
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/352/files#r38870720>.
>



-- 
Best,

Jeff Hinrichs
dundeemt@gmail.com
402.218.1473